src/caches/beliefpropagationcache.jl

```diff
@@ -0,0 +1,221 @@
+default_mt_constructor(inds_e) = ITensor[denseblocks(delta(inds_e))]
+default_mt_storage(ptn::PartitionedGraph) = Dict()
+function default_contractor(contract_list::Vector{ITensor}; kwargs...)
+  return contract_exact(contract_list; kwargs...)
+end
+default_contractor_kwargs() = (; normalize=true, contraction_sequence_alg="optimal")
+@traitfn default_bp_niters(g::::(!IsDirected)) = is_tree(g) ? 1 : nothing
+@traitfn function default_bp_niters(g::::IsDirected)
+  return default_bp_niters(undirected_graph(underlying_graph(g)))
+end
+function message_tensor_diff(mt_a::Vector{ITensor}, mt_b::Vector{ITensor})
+  LHS, RHS = contract(mt_a), contract(mt_b)
+  return 0.5 *
+         norm((denseblocks(LHS) / sum(diag(LHS))) - (denseblocks(RHS) / sum(diag(RHS))))
+end
+
+struct BeliefPropagationCache{PTN,MTS}
+  partitioned_itensornetwork::PTN
+  message_tensors::MTS
+end
+
+#Constructors...
+function BeliefPropagationCache(ptn::PartitionedGraph; mt_storage=default_mt_storage)
+  mts = default_mt_storage(ptn)
+  return BeliefPropagationCache(ptn, mts)
+end
+
+function BeliefPropagationCache(tn::ITensorNetwork, partitioned_vertices; kwargs...)
+  ptn = PartitionedGraph(tn, partitioned_vertices)
+  return BeliefPropagationCache(ptn; kwargs...)
+end
+
+partitioned_itensornetwork(bpc::BeliefPropagationCache) = bpc.partitioned_itensornetwork
+message_tensors(bpc::BeliefPropagationCache) = bpc.message_tensors
+function tensornetwork(bpc::BeliefPropagationCache)
+  return unpartitioned_graph(partitioned_itensornetwork(bpc))
+end
+function NamedGraphs.partitioned_graph(bpc::BeliefPropagationCache)
+  return partitioned_graph(partitioned_itensornetwork(bpc))
+end
+
+function initial_message_tensor(
+  bpc::BeliefPropagationCache, edge::PartitionEdge; mt_constructor=default_mt_constructor
+)
+  ptn = partitioned_itensornetwork(bpc)
+  src_e_itn = subgraph(ptn, src(edge))
+  dst_e_itn = subgraph(ptn, dst(edge))
+  inds_e = commoninds(src_e_itn, dst_e_itn)
+  return mt_constructor(inds_e)
+end
+
+function message_tensor(
+  bpc::BeliefPropagationCache, edge::PartitionEdge; mt_constructor=default_mt_constructor
+)
+  mts = message_tensors(bpc)
+  return haskey(mts, edge) ? mts[edge] : initial_message_tensor(bpc, edge; mt_constructor)
+end
+function message_tensors(
+  bpc::BeliefPropagationCache, edges::Vector{PartitionEdge}; kwargs...
```

Comment:
```suggestion
  bpc::BeliefPropagationCache, edges::Vector{<:PartitionEdge}; kwargs...
```
I would be surprised if it was working the way you wrote it, maybe needs a test.

Reply:
Good spot, I have changed some internal code to directly call this function now anyway so it is being used.